File: codewars/permute_a_palindrome.py
# solution
def permute_a_palindrome(input):
    char_count = {}
    count_odd_char = 0

    for char in input:
        if char in char_count:
            char_count[char] += 1
        else:
            char_count[char] = 1
            count_odd_char = (char_count[char]) % 2 - (char_count[char] - 1) % 2
    return count_odd_char <= 1


# solution  (2)
def permute_a_palindrome(input):
    return sum([1 for i in set(input) if input.count(i) % 2]) < 2

# print(permute_a_palindrome('tactcoa')) # True

# easy solution
def permute_a_palindrome(input):
    flag = 0
    for i in set(input):
        print(input)
        if input.count(i) % 2 == 1:
            print("flag no increment: ", flag)
            flag += 1
            print("flag + 1: ", flag)
            print("i", i )
            print("input count", input.count(i))
    return flag < 2

print(permute_a_palindrome("tactcoa"))

File: codewars/permute a palindrome.js
function permuteAPalindrome(input) {
  let count = 0; // count variable to count the number of odd characters
  const sum = input.split``.reduce(
    // reduce function to count the number of each character in the string
    (a, b) => ((a[b] = a[b] ? a[b] + 1 : 1), a),
    // line above - initial value of a is an empty object {} which is used to store the count of each character
    // ((a[b] = a[b] ? a[b] + 1 : 1), a) - this line is used to count the number of each character 
    {} // initial value of a is an empty object
  );
  for (let i in sum) { // loop through the object to count the number of odd characters
    if (sum[i] % 2 !== 0) { // if the count of the character is odd in the object
      count++; // increment the count variable
    }
    if (count === 2) return false; // if the count is 2, return false
    if (count > 1) return false;
    if (count === 1 && count === 0) return true;
  }
  return true; // if the count is 1 or 0, return true
}

console.log(permuteAPalindrome("Tact Coa")); // true
console.log(permuteAPalindrome("Tact Coaa")); // false
